README.rst
Django FileBrowser
==================

**Media-Management**. (based on https://github.com/sehmaschine/django-filebrowser)

The FileBrowser is an extension to the `Django <http://www.djangoproject.com>`_ administration interface in order to:

* browse directories on your server and upload/delete/edit/rename files.
* include images/documents to your models/database using the ``FileBrowseField``.
* select images/documents with TinyMCE.

Requirements
------------

FileBrowser 3.6 requires

* Django 1.4/1.5/1.6/1.7/1.8/1.9 (http://www.djangoproject.com)
* Pillow (https://github.com/python-imaging/Pillow)

No Grappelli
-----------

This fork removes the dependency on Grappelli.

.. figure:: docs/_static/Screenshot.png
   :scale: 50 %
   :alt: django filebrowser no grappelli

Installation
------------

Latest version:

    pip install -e git+git://github.com/smacker/django-filebrowser-no-grappelli.git#egg=django-filebrowser

Stable version:

    pip install django-filebrowser-no-grappelli

Documentation
-------------

http://readthedocs.org/docs/django-filebrowser/

It also has fake model to show filebrowser in admin dashboard, but you can disable it by setting ``FILEBROWSER_SHOW_IN_DASHBOARD = False``.

Translation
-----------

https://www.transifex.com/projects/p/django-filebrowser/

Releases
--------

* FileBrowser 3.6.2 (Development Version, not yet released, master branch)
* FileBrowser 3.6.1 (November 11th, 2015): Compatible with Django 1.4/1.5/1.6/1.7/1.8

Older versions are available at GitHub, but are not supported anymore.


docs/changelog.rst
:tocdepth: 1

.. |grappelli| replace:: Grappelli
.. |filebrowser| replace:: FileBrowser

.. _changelog:

Changelog
=========

3.6.2 (not yet released)
------------------------

* New: Compatibility with Django 1.9
* New: Better test coverage with tox.
* New: Improved tests.
* New: Add support for using filebrowser from tiny-mce v4
* New: Breadcrumbs on pop-ups
* Fixed: The block usertools not displayed on django 1.8
* Fixed: Fix url tag for django 1.4
* Fixed: Django styles for experimental FileBrowseUploadField

3.6.1 (September 9th, 2015)
---------------------------

* Compatibility with Django 1.8

3.5.8 (November 8th 2015)
--------------------------

* New: Sort by multiple attributes.
* Improved: Add back button from upload page.
* Improved: Show url after upload
* Improved: Update video extensions for common HTML5 video types.
* Improved: Improved Python 3 compatibility.
* Improved: Version size with crop.
* Improved: Fix cannot write mode P as JPEG.
* Fixed: Action pulldown with all documents (not only images).
* Fixed: Version scaling with fixed width and auto height (added tests as well).
* Fixed: Compatibility with Django 1.4.
* Fixed: Management command when generating all versions (fb_version_generate).
* Fixed: Home link with breadcrumbs.
* Fixed: Removed cycle templatetag (for compatibility with Django 1.4 to 1.7).

3.5.7 (December 14th 2014)
---------------------------

* New: Compatibility with Django 1.7.
* Improved: Updated tests because of the new random suffix with get_available_name (django storage).
* Improved: Added an icon in order to mark finished uploads.
* Improved: Show resulting filename (e.g. with suffix, converted) after successful upload.
* Fixed: Permissions with file upload.
* Fixed: Unified json response with _upload_file (no matter if file already exists or not).

3.5.6 (August 16th, 2014)
------------------------

* New: Add fake model to show filebrowser in admin dashboard
* New: Make fake model optional
* Improved: Restore the filebrowser icon for file fields.
* Improved: Escape extension dots in default EXCLUDE list, avoid empty entry
* Improved: Set xframe_options_sameorigin policy on filebrowser views.
* Fixed: Fix filebrowser button for newly added inlines

3.5.5 (April 13th, 2014)
------------------------

* New: Added client-side (JavaScript) file extension validation to the AJAX uploader.
* New: Added experimental Python 3.3 support.
* Improved: Tests with convert/normalize (removed special chars from test filename).
* Fixed: File selection after using search box (CKEditor).
* Fixed: Removed encoding of file URIs with CKEditor.

3.5.4 (February 21st, 2014)
---------------------------

* Fixed: Placeholder functionality (including tests).
* Fixed: Convert/normalize filenames (including tests).
* Fixed: Handling uppercase extensions with browse.

3.5.3 (January 7, 2014)
-----------------------

* New: added ``path_full`` to FileObject.
* Improved: added docx to EXTENSIONS.
* Improved: Recommend pillow instead of PIL as a requirement.
* Improved: Added additional test cases.
* Improved: Updated documentation.
* Improved: Consistent use of storage (e.g. storage.location, storage.url).
* Improved: Removed unnecessary functions (e.g. url_join, url_strip).
* Improved: Moved sort_by_attr to FileListing.
* Improved: Regex matches with file versions on browse.
* Improved: Using django.conf.urls (with django.conf.urls.defaults as fallback).
* Improved: Adding CONTRIBUTING.rst.
* Improved: Removed static Media inner class with fields.
* Improved: Removed search icon with fields (has not being used).
* Improved: Added custom class attributes with filebrowser field.
* Improved: Updated translations.
* Fixed: fixed exception handling with python 2.5.
* Fixed: fixes dir with SEARCH_TRAVERSE true and version select.
* Fixed: Make Django FileUploadHandlers work (also fixed a memory leak).
* Fixed: return correct filename with ``OVERWRITE_EXISTING``.
* Fixed: fb_version_generate with ``FILEBROWSER_VERSIONS_BASEDIR``.
* Fixed: Table sorting with asc/desc.
* Fixed: Added DeprecationWarning for ``FileObject.directory`` and ``FileObject.folder``.

3.5.2 (February 22, 2013)
-------------------------

* Fixed: Use placeholder with version_generate (not only templatetags).
* Fixed: translate extension group name in upload form.
* Fixed: updated filter dropdown HTML.
* Fixed: Make setup.py work with Python 3.
* Fixed: File submit with search traversal.
* Fixed: Fixed fileobject path with Windows.
* Improved: Throwing an exception when in DEBUG and version is not generated (with using the templatetag).
* Compatibility with Django 1.5.

3.5.1 (November 09, 2012)
-------------------------

* Fixed: Documentation with Signals.
* Fixed: File Upload using basic submission.
* Fixed: Added site instance to Signals.
* Improved: Don't hide errors during generate-command.
* Improved: Follow symlinks with generate-command.
* Improved: Added some translations (e.g. for "Upload File").
* New: Setting OVERWRITE_EXISTING.
* New: Added file ``signals.py``.
* New: Support for Django 1.5.

3.5.0 (July 20, 2012)
---------------------

* Compatibility with Django 1.4 and Grappelli 2.4.

3.4.3 (20.6.2012)
-----------------

* Fixed a bug with versions not being generated (in case of capitalized extensions).

3.4.2 (26.3.2012)
-----------------

* Fixed security bug: added staff_member_required decorator to the upload-function.
* Fixed a XSS vulnerability with fb_tags.

3.4.1 (7.3.2012)
----------------

* Fixed an error with quotes (french translation) in upload.html.
* Updated translations.
* FileObject now returns path (with __unicode__ and __str__), instead of filename. This is needed because otherwise form.has_changed will always be triggerde when using a FileBrowseField.
* Fixed a bug with versions and "f referenced before assignment" (e.g. when an image is being deleted)

filebrowser/settings.py
# coding: utf-8

# django imports
from django.conf import settings
from django.utils.translation import ugettext_lazy as _
import re

# PATH AND URL SETTINGS

# Main Media Settings
# WARNING: FILEBROWSER_MEDIA_ROOT and FILEBROWSER_MEDIA_URL will be removed with Filebrowser 3.6.
# Read the documentation on FileBrowser's storages (http://readthedocs.org/docs/django-filebrowser/en/latest/file_storages.html)
MEDIA_ROOT = getattr(settings, "FILEBROWSER_MEDIA_ROOT", settings.MEDIA_ROOT)
MEDIA_URL = getattr(settings, "FILEBROWSER_MEDIA_URL", settings.MEDIA_URL)
# Main FileBrowser Directory. This has to be a directory within MEDIA_ROOT.
# Leave empty in order to browse all files under MEDIA_ROOT.
# DO NOT USE A SLASH AT THE BEGINNING, DO NOT FORGET THE TRAILING SLASH AT THE END.
DIRECTORY = getattr(settings, "FILEBROWSER_DIRECTORY", 'uploads/')

# EXTENSIONS AND FORMATS
# Allowed Extensions for File Upload. Lower case is important.
EXTENSIONS = getattr(settings, "FILEBROWSER_EXTENSIONS", {
    'Folder': [''],
    'Image': ['.jpg', '.jpeg', '.gif', '.png', '.tif', '.tiff'],
    'Document': ['.pdf', '.doc', '.rtf', '.txt', '.xls', '.csv', '.docx'],
    'Video': ['.mov', '.wmv', '.mpeg', '.mpg', '.avi', '.rm'],
    'Audio': ['.mp3', '.mp4', '.wav', '.aiff', '.midi', '.m4p']
})
# Define different formats for allowed selections.
# This has to be a subset of EXTENSIONS.
# e.g., add ?type=image to the browse-URL ...
SELECT_FORMATS = getattr(settings, "FILEBROWSER_SELECT_FORMATS", {
    'file': ['Folder', 'Image', 'Document', 'Video', 'Audio'],
    'image': ['Image'],
    'document': ['Document'],
    'media': ['Video', 'Audio'],
})

# VERSIONS

# Directory to Save Image Versions (and Thumbnails). Relative to site.storage.location.
# If no directory is given, versions are stored within the Image directory.
# VERSION URL: VERSIONS_BASEDIR/original_path/originalfilename_versionsuffix.extension
VERSIONS_BASEDIR = getattr(settings, 'FILEBROWSER_VERSIONS_BASEDIR', '')
# Versions Format. Available Attributes: verbose_name, width, height, opts
VERSIONS = getattr(settings, "FILEBROWSER_VERSIONS", {
    'admin_thumbnail': {'verbose_name': 'Admin Thumbnail', 'width': 60, 'height': 60, 'opts': 'crop'},
    'thumbnail': {'verbose_name': 'Thumbnail (1 col)', 'width': 60, 'height': 60, 'opts': 'crop'},
    'small': {'verbose_name': 'Small (2 col)', 'width': 140, 'height': '', 'opts': ''},
    'medium': {'verbose_name': 'Medium (4col )', 'width': 300, 'height': '', 'opts': ''},
    'big': {'verbose_name': 'Big (6 col)', 'width': 460, 'height': '', 'opts': ''},
    'large': {'verbose_name': 'Large (8 col)', 'width': 680, 'height': '', 'opts': ''},
})
# Quality of saved versions
VERSION_QUALITY = getattr(settings, 'FILEBROWSER_VERSION_QUALITY', 90)
# Versions available within the Admin-Interface.
ADMIN_VERSIONS = getattr(settings, 'FILEBROWSER_ADMIN_VERSIONS', ['thumbnail', 'small', 'medium', 'big', 'large'])
# Which Version should be used as Admin-thumbnail.
ADMIN_THUMBNAIL = getattr(settings, 'FILEBROWSER_ADMIN_THUMBNAIL', 'admin_thumbnail')

# PLACEHOLDER

# Path to placeholder image (relative to storage location)
PLACEHOLDER = getattr(settings, "FILEBROWSER_PLACEHOLDER", "")
# Show Placeholder if the original image does not exist
SHOW_PLACEHOLDER = getattr(settings, "FILEBROWSER_SHOW_PLACEHOLDER", False)
# Always show placeholder (even if the original image exists)
FORCE_PLACEHOLDER = getattr(settings, "FILEBROWSER_FORCE_PLACEHOLDER", False)

# EXTRA SETTINGS

# If set to True, the FileBrowser will not try to import a mis-installed PIL.
STRICT_PIL = getattr(settings, 'FILEBROWSER_STRICT_PIL', False)
# PIL's Error "Suspension not allowed here" work around:
# s. http://mail.python.org/pipermail/image-sig/1999-August/000816.html
IMAGE_MAXBLOCK = getattr(settings, 'FILEBROWSER_IMAGE_MAXBLOCK', 1024*1024)
# Exclude files matching any of the following regular expressions
# Default is to exclude 'thumbnail' style naming of image-thumbnails.
EXTENSION_LIST = []
for exts in EXTENSIONS.values():
    EXTENSION_LIST += [re.escape(ext) for ext in exts if ext]
EXCLUDE = getattr(settings, 'FILEBROWSER_EXCLUDE', (r'_(%(exts)s)_.*_q\d{1,3}\.(%(exts)s)' % {'exts': ('|'.join(EXTENSION_LIST))},))
# Max. Upload Size in Bytes.
MAX_UPLOAD_SIZE = getattr(settings, "FILEBROWSER_MAX_UPLOAD_SIZE", 10485760)
# Normalize filename and remove all non-alphanumeric characters
# except for underscores, spaces & dashes.
NORMALIZE_FILENAME = getattr(settings, "FILEBROWSER_NORMALIZE_FILENAME", False)
# Convert Filename (replace spaces and convert to lowercase)
CONVERT_FILENAME = getattr(settings, "FILEBROWSER_CONVERT_FILENAME", True)
# Max. Entries per Page
# Loading a Sever-Directory with lots of files might take a while
# Use this setting to limit the items shown
LIST_PER_PAGE = getattr(settings, "FILEBROWSER_LIST_PER_PAGE", 50)
# Default Sorting
# Options: date, filesize, filename_lower, filetype_checked
DEFAULT_SORTING_BY = getattr(settings, "FILEBROWSER_DEFAULT_SORTING_BY", "date")
# Sorting Order: asc, desc
DEFAULT_SORTING_ORDER = getattr(settings, "FILEBROWSER_DEFAULT_SORTING_ORDER", "desc")
# regex to clean dir names before creation
FOLDER_REGEX = getattr(settings, "FILEBROWSER_FOLDER_REGEX", r'^[\w._\ /-]+$')
# Traverse directories when searching
SEARCH_TRAVERSE = getattr(settings, "FILEBROWSER_SEARCH_TRAVERSE", False)
# Default Upload and Version Permissions
DEFAULT_PERMISSIONS = getattr(settings, "FILEBROWSER_DEFAULT_PERMISSIONS", 0o755)
# Overwrite existing files on upload
OVERWRITE_EXISTING = getattr(settings, "FILEBROWSER_OVERWRITE_EXISTING", True)

# EXTRA TRANSLATION STRINGS

# The following strings are not availabe within views or templates
_('Folder')
_('Image')
_('Video')
_('Document')
_('Audio')
